### ISSUE TITLE
In supporting of code reuse

### ISSUE THREAD

**Issue by suntong** [gopherjs/gopherjs#751]
I know gopherjs is mainly used/built to run Go in browsers, however, I'm thinking to use it as just a transpiler from Go to Javascript. Hope this does not stretch its purpose too far. 

This brings a problem to me -- for this single line of code, 

 	rand.Seed(time.Now().Unix())

I'm pulling in two packages, `math/rand` & `time`. These too big packages are pulled in just so that I can use a random number in my code. However I believe Javascript has it own way of doing randomization. I.e., a direct translate would not be ideal. Same for using `fmt.Sprintf("%d")` to convert an `int` to `string` because I don't want to [pull in and use `fmt`](http://legacytotheedge.blogspot.ca/2014/03/gopherjs-go-to-javascript-transpiler.html) as Javascript does that with zero cost:

> 1470 LOC and 45kb, uncompressed and unminified.  The bulk of which is the builtin library.
It will only compile what it needs to.  ...  This goes for core packages too.  If I change that code so it requires "fmt", the result explodes to 12845 LOC and 624kb ("fmt" imports a LOT of stuff).

In summary, my question is, what's the best approach so that I can keep my Go lib serving for my Go code, _**while**_ able to transpile it into Javascript with minimum overhead? thx.

**Comment by flimzy** [comment]
I don't really understand the question.

First, if you're not executing in the browser, where do you intend to execute code compiled by GopherJS? Understanding your use-case will probably help understand the rest of your question.

Second, your final question confuses me a bit. I think maybe you're asking how to best write isomorphic code? That is, code that can be run in a (normal) Go runtime environment, and also in a GopherJS transpired environment. Am I on the right path with this understanding?

**Comment by suntong** [comment]
Yeah, totally understand that it is not anything near a normal use case. Your paraphrasing it being _isomorphic code_ is exactly what I was trying to do. Having looking into that myself, I think it might not even possible. 

That being said, I did make some progress/breakthrough. The "_common_" code that I want to share between Go and Javascript is an algorithm, https://github.com/suntong/game24, which is based on http://rosettacode.org/wiki/24_game/Solve#Go. The reason that I do not want to use the http://rosettacode.org/wiki/24_game/Solve#JavaScript version is, on analyzing the JavaScript, I found it ultra-stupid -- to summarize the JavaScript's  algorithm in one sentence, it is making 100000 _random guesses_, and hope one of them is correct. Anyway, I'm not a JavaScript programmer so I can't translate my Go code to JavaScript directly, hence I'm looking into gopherjs. 

Long story short, I 

- [made these changes in order to use gopherjs](https://github.com/suntong/game24/commit/5c3b36ddbe9a8f4911d0055454d48ce70967da7d), then
- added a _fake_ main.go under https://github.com/suntong/game24/tree/master/_other-demo/tojs, 
- generate a .js file using gopherjs
- manually tweak the .js file to [make it want I want](https://github.com/suntong/game24/blob/master/_other-demo/tojs/main.js) (yeah, here is the _isomorphic code_ part, the main.go has to be _fake_, and need extra manual tweaking)
- and finally _run it under nodejs_, which is where I intend to execute code compiled by GopherJS.

and the result is exactly as my Go's output:

```
$ node main.js 

 7 7
 4 9
- XX


 1 2
 7 2
- !!
: (7 - 1) * 2 * 2


 5 7
 5 1
- XX


 8 7
 1 5
- !!
: (8 - 5) * (7 + 1)
```

Again, I understand that it is not anything near a normal use case. So my current question is, 

- do you think if there is any way to make this process any way simpler?
- I ultimate goal is not just run it under nodejs, but to `require` it as a module, so that I can use the `Solve` & `Play` functions defined in it. But I don't understand the generated code structure well enough  to make the change. Could you help a bit here please? 

Thanks!

**Comment by pallavagarwal07** [comment]
@suntong I think you need to do something like this:
https://github.com/pallavagarwal07/gophernet

Basically, I needed a way to compile my go programs with both natively and in the browser, so I created a gophernet library to replace calls of `net/http` with gophernet and the conditional compiler takes care of the rest.
You need to create a wrapper around the rand library. And replace calls of rand with your wrapper library.

**Comment by suntong** [comment]
Thanks @pallavagarwal07, 

where can I read more about _"the conditional compiler"_ please? 
I was looking for a way to transpile Go file by file, and without adding any dependencies supporting codes. 
 
> ... but they are compiled to xhr requests when gopherjs is used. 

> GopherNet provides a way to do HTTP requests from go that get compiled to javascript (using XHR) on using gopherjs by using proper build tags inside the source file(s).

where can I read more about these _"build tags inside the source file"_ as well please? 

thx!

**Comment by pallavagarwal07** [comment]
@suntong 
https://dave.cheney.net/2013/10/12/how-to-use-conditional-compilation-with-the-go-build-tool

https://github.com/gopherjs/gopherjs#general
> The GOARCH value of GopherJS is "js". You may use it as a build constraint: // +build js.

And in the code:
https://github.com/pallavagarwal07/gophernet/blob/8c2cfdfef6ce050bcc1188db8fde7394a1c7f4be/netdef.go#L1
https://github.com/pallavagarwal07/gophernet/blob/8c2cfdfef6ce050bcc1188db8fde7394a1c7f4be/netjs.go#L1

(notice the build tags on line 1)

**Comment by suntong** [comment]
Make perfect sense. Thanks @pallavagarwal07!
